Search user profile

trashstackdev opened the pull request:
# Pull Request


## Trello Ticket

- [Trello Ticket Link](insert link here)
  <!-- Link to the corresponding Trello ticket for better context -->

## Description

- GET user/{username}
- GET user/?query
- POST subscription
- DELETE subscription/{subscriptionId}

## Additional Remarks

Einige Verbesserungsvorschläge in der Readme

NilsAbels committed:
first commit

trashstackdev committed:
Refactor setupMocks function in router_test.go

The setupMocks function now takes a testing.T parameter which it uses to set environment variables. This adjustment to the setupMocks function facilitates improved handling of cryptographic keys within the testing environment. Additionally, the error handling mechanism for failed ed25519 key generation is updated from fatal to error logging for better error management. The changes include the addition of the testing.T parameter in the setupMocks function call across all test cases.

trashstackdev committed:
Improve JWT token handling and context key usage

Added a check for JWT token existence in the authorization header of incoming requests in JWTMiddleware. Replaced hard-coded "claims" key with an util-defined key in JWT handling and context retrieval in handlers. Newly introduced a utils file, contextkeys.go to contain context keys, improving usage consistency across handlers.

trashstackdev committed:
Improve JWT key generation and storage

Updated the JWTManager to automatically generate and store keys in PEM format if they are not already present, enhancing security and usability. Errors during key reading, decoding, and parsing are now logged for potential troubleshooting.

trashstackdev committed:
Modify user handlers, initialize JWT manager with keys from file

User handlers were corrected for proper SQL query formation. The JWTManager initialization methods have been updated; the app now loads the key pair either from provided keys or from file for enhanced flexibility and ease of use. Modified initializer.go and router_test.go to fit these new changes in functions. Added Docker support for reading the key files directly in docker-compose.yml.

trashstackdev committed:
Refactor SQL schema and update SQL queries in user handlers

Updated SQL schema to modify database constraints, table names, and column properties for increased data integrity and improved naming convention. Also, adjusted SQL queries in the user handlers for alignment with the updated schema, ensuring correct working of user actions. Changes reflect in database model, server_alpha_db.sql and user_handler.go files.

trashstackdev committed:
Add subscription functionality to user handlers

Implemented subscription functionality to the user handlers, allowing users to subscribe and unsubscribe to other users. Additionally, refactored the GetUser method in the UserHandler to improve readability and performance. Subscription routes were also added to the router. New schemas were created for the SubscriptionRequest and SubscriptionDTO.

trashstackdev committed:
Add API changes proposals to README

trashstackdev committed:
Refactor subscription validation in user handler

The main change reduces the complexity of the subscription validation process in the user handler. The code has been simplified by removing the helper function 'checkSubscriptionExistence'. This function was replaced by a direct database query, thus improving readability and potentially enhancing performance through the elimination of the extra function call.

trashstackdev committed:
Implement user search feature and revise database schema

Implemented a comprehensive search feature in the user handler that allows user search using query parameters, offset, and limit. This leverages on the fuzzystrmatch SQL extension for pattern-matching capabilities. Database schema changes include the addition of new constraints and modifications of column structures to accommodate this new functionality.

krieger2501 commented on README.md:
```diff
@@ -87,5 +87,43 @@ The service can be configured via environment variables. The following variables
+# Verbesserungsvorschläge/Änderungen
+
+## 1 User
+
+- [ ] **ALLE** API Ressourcen umbenennen (z.B. `users` -> `user`)
+
+### Subscriptions
+- [ ] `GET /users/:user/subscribers` -> Alle Subscriber, die dem User mit der ID `:user` folgen
+
+- [ ] `GET /users/:user/subscriptions` -> Alle User, den der User mit der ID `:user` folgt
+- [ ] `POST /users/:user/subscriptions` -> User mit der ID `:user` abonnieren
+- [ ] `DELETE /users/:user/subscriptions` -> User mit der ID `:user` nicht mehr abonnieren
+
+## 2 Posts
+- [ ] `POST   /posts/` -> Post erstellen
+- [ ] `GET    /posts/:postid` -> Post mit der ID `:postid`
+- [ ] `PUT    /posts/:postid` -> Post mit der ID `:postid` aktualisieren
+- [ ] `DELETE /posts/:postid` -> Post mit der ID `:postid` löschen
+
+### 2.1 Feed
+- [ ] `GET /feed/private` -> Privater Feed des Users
+- [ ] `GET /feed/public` -> Öffentlicher Feed von allen Posts
+
+
+### 2.2 Comments
+- [ ] `GET /posts/:postid/comments` -> Alle Kommentare zu Post mit der ID `:postid`
+- [ ] `POST /posts/:postid/comments` -> Kommentar zu Post mit der ID `:postid` hinzufügen
+- [ ] `GET /posts/:postid/comments/:commentid` -> Kommentar mit der ID `:commentid` abrufen
+- [ ] `PUT /posts/:postid/comments/:commentid` -> Kommentar mit der ID `:commentid` aktualisieren
+- [ ] `DELETE /posts/:postid/comments/:commentid` -> Kommentar mit der ID `:commentid` löschen
+
+### 2.3 Likes
+- Posts
+  - [ ] `GET /posts/:postid/likes` -> Alle Likes zu Post mit der ID `:postid` (Gibt eine Liste mit Usern zurück)
+  - [ ] `POST /posts/:postid/likes` -> Like zu Post mit der ID `:postid` hinzufügen
+  - [ ] `DELETE /posts/:postid/likes` -> Like zu Post mit der ID `:postid` entfernen
+- Comments
+  - [ ] `GET /posts/:postid/comments/:commentid/likes` -> Alle Likes zu Kommentar mit der ID `:commentid` (Gibt eine Liste mit Usern zurück)
+  - [ ] `POST /posts/:postid/comments/:commentid/likes` -> Like zu Kommentar mit der ID `:commentid` hinzufügen
+  - [ ] `DELETE /posts/:postid/comments/:commentid/likes` -> Like zu Kommentar mit der ID `:commentid` entfernen
```
Falscher Ort dafür, oder? Kannst das unter nem separaten /api ordner machen

trashstackdev replied:
Ich mache das glaub komplett aus dem repo raus, braucht nur ne ablage

krieger2501 commented on deployments/server_alpha_db.dbm:
unterschied zur `.sql`-datei?

trashstackdev replied:
.dbm ist ein xml die von Porgrammen wie pgModeler gelesen werden kann, falls jemand dort das Schema anpassen will

krieger2501 commented on internal/routing/router.go:
bei meinem nächsten ticket refactor ich das in authorized und unauthorized routes

krieger2501 commented on internal/routing/handlers/user_handler.go:
```diff
@@ -25,6 +27,10 @@ type UserHdl interface {
 	ActivateUser(w http.ResponseWriter, r *http.Request)
 	ResendToken(w http.ResponseWriter, r *http.Request)
 	LoginUser(w http.ResponseWriter, r *http.Request)
+	HandleGetUserRequest(w http.ResponseWriter, r *http.Request)
```
Der Name passt iwie nicht so zum Rest der Methoden, vlt. eher sowas wie `GetUser`?

trashstackdev replied:
DeepSource heult rum dass laut Go idiome ein "Get..." darauf hindeutet, dass das User Objekt returned wird, was eigentlich nicht passiert

krieger2501 replied:
ja alright, fair enough

krieger2501 commented on internal/routing/handlers/user_handler.go:
```diff
@@ -437,3 +443,308 @@ func generateToken() string {
 }
+
+func (handler *UserHandler) HandleGetUserRequest(w http.ResponseWriter, r *http.Request) {
+	ctx, cancel := context.WithDeadline(r.Context(), time.Now().Add(500*time.Second))
```
500 Sekunden? that's a lot, da braucht man auch keine deadline mehr, lmao

trashstackdev replied:
Hat sich mein Debug eingeschlichen 😙

krieger2501 commented on internal/routing/handlers/user_handler.go:
```diff
@@ -437,3 +443,308 @@ func generateToken() string {
+	var userId uuid.UUID
+
+	// Get username from path
+	username := chi.URLParam(r, "username")
```
Jetzt hast du schon einen ClaimsKey, da können wir auch URLKeys machen, oder?

krieger2501 commented on internal/routing/handlers/user_handler.go:
```diff
@@ -437,3 +443,308 @@ func generateToken() string {
+	jwtUserId := claims["sub"].(string)
+
+	// Get the subscription ID
+	queryString = "SELECT subscription_id FROM alpha_schema.subscriptions WHERE subscriber_id = $1 AND subscribee_id = $2"
+	row = handler.DatabaseManager.GetPool().QueryRow(ctx, queryString, jwtUserId, userId)
+	if err := row.Scan(&user.SubscriptionId); err != nil {
+		user.SubscriptionId = nil
+	}
```
Hier wäre vlt. noch ein primitiver Check sinnvoll, ob es das eigene Profil ist. Dann kann man sich die SQL-Abfrage hier sparen

trashstackdev replied:
Verstehe nicht ganz wieso man diese SQL Query sparen könnte? Man muss doch noch überprüfen ob man den aufgerufen User subscribed oder nicht?

krieger2501 replied:
Wenn es dein eigenes Profil ist, brauchst du das ja nicht. Kannst dich ja schlecht selber subscriben

trashstackdev replied:
Achso mit "sparen" meinst du die SQL nicht auszuführen, wenn wir vorerst überprüfen ob dass der selbe User ist.
Ich hatte jetzt einfach in den selben topf geworfen weil dies implizit der gleiche Check entspricht aber falls du gerne Datenbankabfragen ersparen möchtest kann ich das gerne anpassen und einen weiteren Check einbauen.

trashstackdev replied:
Ich sollte mal anfangen Kommata zu benutzen :)

krieger2501 replied:
Also ist kein muss, das spart denk ich nur minimal Performance. Könnens auch gern so lassen

krieger2501 commented on internal/routing/handlers/user_handler.go:
```diff
@@ -437,3 +443,308 @@ func generateToken() string {
+	}
+
+	// Get the usernames of the users
+	queryString = "SELECT username FROM alpha_schema.users WHERE user_id = $1"
```
Der Call ist hier unnötig, du kannst weiter oben schon den Usernamen zusätzlich zur UserId holen. Für den anderen Usernamen kannst du wieder den JWT nutzen, der steht da auch drin

krieger2501 commented on internal/routing/handlers/user_handler.go:
```diff
@@ -437,3 +443,308 @@ func generateToken() string {
+	jwtUserId := claims["sub"].(string)
+
+	// Get subscriptionId from path
+	subscriptionId := chi.URLParam(r, "subscriptionId")
```
hier evtl. auch ein `URLKey`

krieger2501 commented on internal/routing/handlers/user_handler.go:
```diff
@@ -437,3 +443,308 @@ func generateToken() string {
+	if offset > records {
+		utils.WriteAndLogError(w, schemas.BadRequest, http.StatusBadRequest, errors.New("offset greater than records"))
+		return
+	}
```
Versteh den Check hier nicht ganz? Die `records` in Postman beziehen sich auf die Gesamtheit der Records, die den Filterkriterien entsprechen. Hier schreibst du einfach alle User rein, die wir haben

trashstackdev replied:
Wollte nur verhindern dass jemand den Offset über die Records hinaus setzt und dann keine Ergebnisse mehr hat

krieger2501 commented on internal/routing/handlers/user_handler.go:
```diff
@@ -437,3 +443,308 @@ func generateToken() string {
+	}
+
+	// Get the users that match the search query
+	queryString = "SELECT username, nickname, profile_picture_url, levenshtein(username, $1) as ld FROM alpha_schema.users ORDER BY ld LIMIT $2 OFFSET $3"
```
Funktioniert die PSQL-Funktion einfach so? Muss man beim select nicht irgendwas besonderes angeben

trashstackdev replied:
Ja in der .sql datei erstelle bzw. Importe ich noch die entsprechende Extension "fuzzysearch" die diese Funktion mitliefert

krieger2501 commented on internal/routing/handlers/user_handler.go:
```diff
@@ -437,3 +443,308 @@ func generateToken() string {
+	defer rows.Close()
+
+	// Create a list of users
+	users := make([]schemas.UserSearchDTO, 0)
```
Kannst den Array hier auf `limit` begrenzen, da `limit` ja die Anzahl der gewünschten Records angibt.

krieger2501 commented on internal/routing/handlers/user_handler.go:
```diff
@@ -437,3 +443,308 @@ func generateToken() string {
+			return
+		}
+		users = append(users, user)
+	}
```
Dementsprechend brauchen wir auch nicht alle zutreffenden Records hier, sondern nur von `[offset, offset + limit]`

trashstackdev replied:
Der offset und das limit sind eig schon in der sql query angegeben

krieger2501 replied:
Aber das wollen wir doch nicht, oder? Ich konsolidier mal die und die Konversation unter der, da es das gleiche Thema behandelt.

Was wäre denn eine sinnvolle Levenshtein-Distanz, auf die man sich festlegen kann? Dann kann man das Limit aus der SQL-Query rausnehmen und halt wie eigentlich vorgedacht, `records` richtig befüllen mit der tatsächlichen Anzahl an Treffern.

trashstackdev replied:
### Bezüglich offset/limit in der Query:
Es kommt halt alles auf unsere Erwartungen an. Ich habe jetzt es hier einfach gehalten und gebe vorerst alle User zurück ohne eingrenzung aber limit und offset impliziert in der SQL-Query, weil wir eh nicht mehr als 10 User oder so in der Datenbank haben werden.

Alternativ könnten wir die levenshtein distanz in der Query cappen und dann den offset und limit später anwenden. Ich denke nur dass die beiden eine gleiche Performance haben weil du trotzdem über alle User in der Datenbank eine Levenshtein Distanz ausrechnen musst, bevor du sie dementsprechend. 

### Bezüglich dem Cap:
Ich denke ein Cap von 5 für Usernamen passt aufgrund der kleinen Anzahl an Usern. Wenn du willst kann ich das noch einbauen.

krieger2501 commented on internal/routing/handlers/user_handler.go:
```diff
@@ -437,3 +443,308 @@ func generateToken() string {
+	paginationDto := schemas.Pagination{
+		Offset:  offset,
+		Limit:   limit,
+		Records: records,
```
Records wie oben erwähnt bezieht sich auf alle zutreffenden Records, nicht auf allgemein alle

trashstackdev replied:
War auch mein Plan, aber die Levenstein Geschichte bezieht sich aber immer auf alle Möglichen User. Wir bräuchten noch weitere Parameter um anzugeben, nur die User wiederzugeben die unter 5 Distanz haben oder so ähnlich

krieger2501 reviewed (changes requested):
Sieht soweit gut aus, blockend sind für mich auf jeden Fall die `SearchUser`-Methode und der Call in `Subscribe`. Den Rest kannst du als Anmerkung beachten

krieger2501 committed:
Add CODEOWNERS file

trashstackdev committed:
Refactor user handler and optimize transaction timeouts

Refactored user handler to utilize constants for parameter keys, greatly improving code readability and maintainability. Transaction timeouts were also optimized from 500 seconds to a more reasonable 10 seconds.

krieger2501 commented:
@930C Was wäre hier ein ETA? Will keinen Stress machen, aber hätte gerne die JWT-Changes auf dem Main-branch. Alternativ können wir auch nur die JWT-Changes auf einem separaten Branch schon durchpushen

trashstackdev committed:
Refactor user search

The change refactors the user search feature by moving offset and limit verifications to the list of retrieved users instead of the database query. This approach allows reducing the number of expensive database operations, thereby optimizing performance. The refactoring also eliminates the UserSearchDTO structure which is found to be redundant, making the codebase cleaner and more efficient.

trashstackdev commented:
@Frecherenkel60 habs gepusht

krieger2501 commented on internal/routing/handlers/user_handler.go:
```diff
@@ -437,3 +443,293 @@ func generateToken() string {
+	// Pre-checks on offset and limit
+	if offset > len(users) {
+		offset = len(users)
+	}
```
Sollten wir daraus eventuell einen eigenen Fehlerfall machen?

trashstackdev commented:
@Frecherenkel60 merge ja/nein?

krieger2501 reviewed (approved):
LGTM, mit der kleinen Anmerkung was wir mit dem Fall machen, wenn das Offset zu hoch ist

trashstackdev committed:
Improve offset and limit checks in user search

The user search now checks the offset and limit against the total number of retrieved records instead of the entire user table from the database. If either is invalid, the function immediately writes an error response and exits. This change optimizes performance by potentially reducing the number of database operations, and improves error reporting by providing more specific error messages.

krieger2501 commented on internal/routing/handlers/user_handler.go:
```diff
@@ -437,3 +443,295 @@ func generateToken() string {
+		utils.WriteAndLogError(w, schemas.BadRequest, http.StatusBadRequest, errors.New("offset invalid"))
+		return
+	}
+	end := offset + limit
+	if end > records {
+		utils.WriteAndLogError(w, schemas.BadRequest, http.StatusBadRequest, errors.New("limit invalid"))
+		return
+	}
```
Das würde ich so lassen wie davor haha. Meinte aber auch ob wir einen ganz eigenen Fehlerfall für einen zu hohen Offset haben wollen
